File: Py/P8_DAG.py
class DAG:
    def __init__(self):
        self.i = self.j = self.k = 0
        self.count = -1
        self.str = [None] * 10  # Store input code sequences
        self.table = [[""] * 10 for _ in range(10)]  # Initialize table for DAG

        print("Saail Chavan KFPMSCCS016\n")
        print("Enter the sequence of code:")
        print("Enter 'q' to Quit")

        # Input the sequence of code lines
        for self.i in range(len(self.str)):
            self.str[self.i] = input()
            if self.str[self.i] == "q":
                break
            else:
                self.count += 1

        # Display the sequence of code
        print("\nThe sequence of code is:")
        for line in self.str:
            if line == "q":
                break
            print(line)

    def tablestruct(self):
        # Fill the table based on code
        for self.i in range(self.count + 1):
            if len(self.str[self.i]) == 3:  # Case for D=A
                self.table[self.i][0] = self.str[self.i][0]
                self.table[self.i][1] = self.str[self.i][1]
                self.table[self.i][2] = self.str[self.i][2]
            elif len(self.str[self.i]) == 5:  # Case for A=B+C
                self.table[self.i][0] = self.str[self.i][0]
                self.table[self.i][1] = self.str[self.i][3]
                self.table[self.i][2] = self.str[self.i][2]
                self.table[self.i][3] = self.str[self.i][4]

        # Optimization: Removing redundant computations
        for self.i in range(self.count + 1):
            for self.j in range(self.i + 1, self.count + 1):
                if len(self.str[self.i]) == 5 and len(self.str[self.j]) == 5:
                    if self.str[self.i][2:5] == self.str[self.j][2:5]:
                        self.table[self.i][0] += f",{self.str[self.j][0]}{self.str[self.j][4]}"
                        self.table[self.j] = [""] * 10

        # Handle case for D=A
        for self.i in range(self.count + 1):
            if len(self.str[self.i]) == 3:
                for self.j in range(self.count):
                    if self.str[self.i][2] == self.str[self.j][0]:
                        self.table[self.j][0] += f",{self.str[self.i][0]}"
                        self.table[self.i] = [""] * 10

        # Print the table
        print("\nLabel\tOperator\tLeft\tRight")
        for row in self.table[:self.count + 1]:
            print("\t".join(row[:self.count + 2]))

if __name__ == "__main__":
    dag = DAG()
    dag.tablestruct()

File: OG/P8_DAG.java
import java.io.*;
public class DAG
{
	int i,j,k;
	int count=-1,flag=0;
	String str[]=new String[10];
	String table[][]=new String[10][10];
	public DAG()
	{
		try
		{
		 BufferedReader br=new BufferedReader(new InputStreamReader(System.in)); 
		System.out.println("Saail Chavan KFPMSCCS016\n");
		 System.out.println("Enter The Sequence of code:");
		 System.out.println("Enter q to Quit");
			for(i=0;i<str.length;i++)
			{
			  str[i]=br.readLine();
			 if(str[i].equals("q"))
			 break;
			 else
				 count++;
		
			}
			System.out.println("The Sequence of code are:");
			for(i=0;i<str.length;i++)
			{
		
			if(str[i].equals("q"))
			 break;
			 System.out.println(str[i]);
			}
		}
		catch(IOException e)
		{}
	}
	public void tablestruct()
	{
		for(i=0;i<str.length;i++)
		{
			for(j=0;j<str.length;j++)
			{
				table[i][j]="";
			}
		}
		try
		{
		for(i=0;i<=count;i++)
		{
			for(j=0;j<=count+1;j++)
			{
				if(str[i].length()==3)//D=A
				{
					if(j==0)
					table[i][j]=str[i].substring(0,1);
					if(j==1)
					table[i][j]=str[i].substring(1,2);
					if(j==2)
					table[i][j]=str[i].substring(2);
				}
				if(str[i].length()==5)//A=B+C
				{
					if(j==0)
					table[i][j]=str[i].substring(0,1);
					if(j==1)
					table[i][j]=str[i].substring(3,4);
					if(j==2)
					table[i][j]=str[i].substring(2,3);
					if(j==3)
					table[i][j]=str[i].substring(4);
				}
			}
		}
		}
		catch(NullPointerException e)
		{}
		
		for(i=0;i<=count;i++)
		{
			for(j=i+1;j<=count;j++)
			{
				if(str[i].length()==5 && str[j].length()==5)
				{
					if(str[i].substring(2,5).equals(str[j].substring(2,5)))
					{
					 table[i][0]= table[i][0].concat(","+str[j].substring(0,1)+str[j].substring(4));
					  for(k=0;k<=count;k++)
						 table[j][k]="";
					}
				}
			}
				if(str[i].length()==3)
				{

					if(i==count)
					{
						for(j=count-1;j>=0;j--)
						{
							if(str[i].substring(2,3).equals(str[j].substring(0,1)))
							{
							  table[j][0]= table[j][0].concat(","+str[i].substring(0,1));
							  for(k=0;k<=count;k++)
								table[i][k]="";
							}
						}
					}
					else
					{
						for(j=i+1;j<=count;j++)
						{
							if(str[i].substring(2,3).equals(str[j].substring(0,1)))
							{
							  table[i][0]= table[i][0].concat(","+str[j].substring(0,1));
							  for(k=0;k<=count;k++)
								table[j][k]="";
							}
						}
					}
				}
				
			}
		System.out.println();
		System.out.print("Label"+"      "+"Operator"+"  "+"Left"+"  "+"Right");
		System.out.println();
		for(i=0;i<=count;i++)
		{
			for(j=0;j<=count+1;j++)
			{
				System.out.print(table[i][j]+"\t      ");
			}
			System.out.println();
		}
	}
	public static void main(String arg[])
	{
		DAG d=new DAG();
		d.tablestruct();
	}
}
